examples/proc_macro.rs

```diff
@@ -26,33 +26,40 @@
 
 use jsonrpsee::{
 	proc_macros::rpc,
-	types::{async_trait, error::Error},
+	types::{async_trait, error::Error, Subscription},
 	ws_client::WsClientBuilder,
-	ws_server::{RpcModule, SubscriptionSink, WsServerBuilder},
+	ws_server::{SubscriptionSink, WsServerBuilder},
 };
 use std::net::SocketAddr;
 
+type ExampleHash = [u8; 32];
+type ExampleStorageKey = Vec<u8>;
+
 #[rpc(server, client, namespace = "state")]
-pub trait Rpc {
+pub trait Rpc<Hash: std::fmt::Debug, StorageKey> {
 	/// Async method call example.
-	#[method(name = "getPairs")]
-	async fn storage_pairs(&self, prefix: usize, hash: Option<u128>) -> Result<Vec<usize>, Error>;
+	#[method(name = "getKeys")]
+	async fn storage_keys(&self, storage_key: StorageKey, hash: Option<Hash>) -> Result<Vec<StorageKey>, Error>;
 
-	/// Subscription that take `Option<Vec<u8>>` as input and produces output `Vec<usize>`.
-	#[subscription(name = "subscribeStorage", unsub = "unsubscribeStorage", item = Vec<usize>)]
-	fn subscribe_storage(&self, keys: Option<Vec<u8>>);
+	/// Subscription that takes a `StorageKey` as input and produces a `Vec<Hash>`.
+	#[subscription(name = "subscribeStorage", unsub = "unsubscribeStorage", item = Vec<Hash>)]
+	fn subscribe_storage(&self, keys: Option<Vec<StorageKey>>);
 }
 
 pub struct RpcServerImpl;
 
 #[async_trait]
-impl RpcServer for RpcServerImpl {
-	async fn storage_pairs(&self, _prefix: usize, _hash: Option<u128>) -> Result<Vec<usize>, Error> {
-		Ok(vec![1, 2, 3, 4])
+impl RpcServer<ExampleHash, ExampleStorageKey> for RpcServerImpl {
+	async fn storage_keys(
+		&self,
+		storage_key: ExampleStorageKey,
+		_hash: Option<ExampleHash>,
+	) -> Result<Vec<ExampleStorageKey>, Error> {
+		Ok(vec![storage_key])
 	}
 
-	fn subscribe_storage(&self, mut sink: SubscriptionSink, keys: Option<Vec<u8>>) {
-		sink.send(&keys.unwrap_or_default()).unwrap();
+	fn subscribe_storage(&self, mut sink: SubscriptionSink, _keys: Option<Vec<ExampleStorageKey>>) {
+		sink.send(&vec![[0; 32]]).unwrap();
```

Comment:
technically, we don't have any compile time check that actually only `Vec<Hash>` can sent over the sink because send has T: Serialize bound.... not sure if we want ensure that at compile-time